src/libos/crates/page-cache/src/cached_disk.rs

```diff
@@ -0,0 +1,583 @@
+    /// Read cache content from `offset` into the given buffer.
+    ///
+    /// The length of buffer and offset can be arbitrary.
+    /// On success, return the number of readed bytes.
+    pub async fn read(&self, offset: usize, buf: &mut [u8]) -> Result<usize> {
+        self.0.read(offset, buf).await
+    }
+
+    /// Write buffer content into cache starting from `offset`.
+    ///
+    /// The length of buffer and offset can be arbitrary.
+    /// On success, return the number of written bytes.
+    pub async fn write(&self, offset: usize, buf: &[u8]) -> Result<usize> {
+        self.0.write(offset, buf).await
+    }
+
+    /// Flush all changes onto the backing disk.
+    pub async fn flush(&self) -> Result<usize> {
+        self.0.flush_disk().await
+    }
+
+    /// Flush specific pages onto the backing disk.
+    pub async fn flush_pages(&self, pages: &Vec<BlockId>) -> Result<usize> {
+        self.0.flush_pages(pages).await
+    }
+}
+
+impl<A: PageAlloc> Drop for CachedDisk<A> {
+    fn drop(&mut self) {
+        self.0.is_dropped.store(true, Ordering::Relaxed);
+        self.0.flusher_wq.wake_all();
+    }
+}
+
+/// BlockRange represents block information
+/// of one read/write operation.
+struct BlockRange {
+    // id of first block
+    begin_id: BlockId,
+    // offset of first block
+    begin_offset: usize,
+    // id of last block
+    end_id: BlockId,
+    // offset of last block
+    end_offset: usize,
+}
+
+impl BlockRange {
+    fn from(offset: usize, buf: &[u8]) -> Self {
+        let begin_id = offset / BLOCK_SIZE;
+        let begin_offset = offset % BLOCK_SIZE;
+        let end_id = (offset + buf.len()) / BLOCK_SIZE;
+        let end_offset = (offset + buf.len()) % BLOCK_SIZE;
+        Self {
+            begin_id,
+            begin_offset,
+            end_id,
+            end_offset,
+        }
+    }
+}
+
+impl<A: PageAlloc> Inner<A> {
+    /// Byte-level read function.
+    pub async fn read(&self, offset: usize, buf: &mut [u8]) -> Result<usize> {
+        let block_range = self.check_rw_args(offset, buf);
+
+        let mut buf_pos = 0;
+        let mut read_len = 0;
+        // Read one partial/whole page each time
+        for current_id in block_range.begin_id..=block_range.end_id {
+            let mut begin_pos = block_range.begin_offset;
+            let mut end_pos = block_range.end_offset;
+            if current_id != block_range.begin_id {
+                begin_pos = 0;
+            }
+            if current_id != block_range.end_id {
+                end_pos = BLOCK_SIZE;
+            }
+
+            let read_buf = &mut buf[buf_pos..buf_pos + end_pos - begin_pos];
+            if read_buf.len() > 0 {
+                read_len +=
+                    if let Ok(len) = self.read_one_page(current_id, read_buf, begin_pos).await {
+                        len
+                    } else {
+                        error!("[CachedDisk] read one page failed, id: {}", current_id);
+                        0
```

Comment:
Returning 0 is not enough. The loop must exit! A short read is ok, but not a short read with _holes_.